src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/CodeGeneration/CodeWriterExtensions.cs

```diff
@@ -1027,120 +1034,124 @@ _writer.LastChar is char ch &&
+        }
+
+        public static LinePragmaScope Enhanced(CodeRenderingContext context, SourceSpan span, int characterOffset, bool suppressLineDefaultAndHidden)
+        {
+            var writer = context.CodeWriter;
+            var options = context.Options;
+
+            // Are we allowed to write an enhanced line pragma?
+            if (!options.UseEnhancedLinePragma)
             {
-                WriteLineNumberDirective(writer, span, ensurePathBackslashes);
+                var standardScope = Standard(context, span, suppressLineDefaultAndHidden);
+
+                // If the caller requested an enhanced line directive, but we fell back to
+                // a regular one, write out the extra padding and add a source mapping.
+                writer.WritePadding(offset: 0, span, context);
+                context.AddSourceMappingFor(span);
+
+                return standardScope;
             }
 
-            // Capture the line index after writing the #line directive.
-            _startLineIndex = writer.Location.LineIndex;
+            var (originalIndent, ensurePathBackSlashes) = WritePreamble(writer, options);
 
-            if (useEnhancedLinePragma)
-            {
-                // If the caller requested an enhanced line directive, but we fell back to regular ones, write out the extra padding that is required
-                if (!_context.Options.UseEnhancedLinePragma)
-                {
-                    context.CodeWriter.WritePadding(0, span, context);
-                    characterOffset = 0;
-                }
+            writer.WriteEnhancedLineNumberDirective(span, characterOffset, ensurePathBackSlashes);
+            context.AddSourceMappingFor(span, characterOffset);
+
+            return new(context, span, isEnhanced: true, originalIndent, suppressLineDefaultAndHidden);
+        }
 
-                context.AddSourceMappingFor(span, characterOffset);
+        private static (int originalIndent, bool ensureBackSlashes) WritePreamble(CodeWriter writer, RazorCodeGenerationOptions options)
+        {
+            // Capture the current indent and reset to 0 for writing directives.
+            var originalIndent = writer.CurrentIndent;
+            writer.CurrentIndent = 0;
+
+            // Ensure that we start on a new line.
+            writer.EnsureNewLine();
+
+            if (!options.SuppressNullabilityEnforcement)
+            {
+                writer.WriteLine("#nullable restore");
             }
+
+            var ensurePathBackSlashes = options.RemapLinePragmaPathsOnWindows && PlatformInformation.IsWindows;
+
+            return (originalIndent, ensurePathBackSlashes);
         }
 
         public void Dispose()
         {
-            // Need to add an additional line at the end IF there wasn't one already written.
-            // This is needed to work with the C# editor's handling of #line ...
-            var endsWithNewline = _writer.LastChar is '\n';
+            if (_context is null)
```

Comment:
>if (_context is null) [](http://example.com/codeflow?start=12&length=21)

might be nice to comment this is for the LinePragmaScope.None case as it was a bit confusing until I realized that